File: 0960-delete-columns-to-make-sorted-iii/0960-delete-columns-to-make-sorted-iii.py
class Solution:
    def minDeletionSize(self, strs: List[str]) -> int:
        dp = [1 for _ in range(len(strs[0]))]
        for i in range(len(strs[0]) - 2, -1, -1):
            for j in range(i + 1, len(strs[0])):
                violation = False
                for k in range(len(strs)):
                    if strs[k][i] > strs[k][j]:
                        violation = True
                        break
                if not violation:
                    dp[i] = max(dp[i], dp[j] + 1)
        return len(strs[0]) - max(dp)

File: 0960-delete-columns-to-make-sorted-iii/0960-delete-columns-to-make-sorted-iii.cpp
class Solution {
public:
    int minDeletionSize(vector<string>& strs) {
        int res = 0;
        vector<int> dp(strs[0].size(), 1);
        for(int i = strs[0].size() - 2; i >= 0; i--)
        {
            for(int j = i + 1; j < strs[0].size(); j++)
            {
                bool violation = false;
                for(int k = 0; k < strs.size(); k++)
                {
                    if(strs[k][i] > strs[k][j])
                    {
                        violation = true;
                        break;
                    }
                }
                if(!violation)
                {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        for(int i = 0; i < dp.size(); i++)
        {
            res = max(res, dp[i]);
        }
        return strs[0].size() - res;
    }
};